pkg/logs/input/journald/tailer_systemd.go

```diff
@@ -0,0 +1,194 @@
+		// if no units are defined, collect all the logs of the journal by default.
+		match := sdjournal.SD_JOURNAL_FIELD_SYSTEMD_UNIT + "=" + unit
+		err := t.journal.AddMatch(match)
+		if err != nil {
+			return fmt.Errorf("could not add filter %s: %s", match, err)
+		}
+	}
+
+	blacklist := make(map[string]bool)
+	for _, unit := range config.ExcludeUnits {
+		blacklist[unit] = true
+	}
+	t.blacklist = blacklist
+
+	return nil
+}
+
+// seek seeks to the cursor if it is not empty or the end of the journal,
+// returns an error if the operation failed.
+func (t *Tailer) seek(cursor string) error {
+	if cursor != "" {
+		err := t.journal.SeekCursor(cursor)
+		if err != nil {
+			return err
+		}
+		// must skip one entry since the cursor points to the last committed one.
+		_, err = t.journal.NextSkip(1)
+		return err
+	} else {
+		return t.journal.SeekTail()
+	}
+}
+
+// tail tails the journal until a message stop is received.
+func (t *Tailer) tail() {
+	defer func() {
+		t.journal.Close()
+		t.done <- struct{}{}
+	}()
+	for {
+		select {
+		case <-t.stop:
+			// stop tailing journal
+			return
+		default:
+			n, err := t.journal.Next()
+			if err != nil && err != io.EOF {
+				t.errHandler.Handle(NewTailError(t.Identifier(), err))
+				return
+			}
+			if n < 1 {
+				// no new entry
+				t.journal.Wait(defaultWaitDuration)
+				continue
+			}
+			entry, err := t.journal.GetEntry()
+			if err != nil {
+				log.Warnf("Could not retrieve journal entry: %s", err)
+				continue
+			}
+			if t.shouldDrop(entry) {
+				continue
+			}
+			t.outputChan <- t.toMessage(entry)
+		}
+	}
+}
+
+// shouldDrop returns true if the entry should be forwarded,
+// returns false otherwise.
+func (t *Tailer) shouldDrop(entry *sdjournal.JournalEntry) bool {
+	unit, exists := entry.Fields[sdjournal.SD_JOURNAL_FIELD_SYSTEMD_UNIT]
+	if !exists {
+		return false
+	}
+	if _, blacklisted := t.blacklist[unit]; blacklisted {
+		// drop the entry
+		return true
+	}
+	return false
+}
+
+// toMessage transforms a journal entry into a message.
+// A journal entry has different fields that may vary depending on its nature,
+// for more information, see https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html.
+func (t *Tailer) toMessage(entry *sdjournal.JournalEntry) message.Message {
+	origin := message.NewOrigin(t.source)
+	origin.Identifier = t.Identifier()
+	origin.Cursor, _ = t.journal.GetCursor()
+	origin.Source = journaldIntegration
+	origin.Service = t.getService(entry)
+	return message.New(t.getContent(entry), origin, nil)
+}
+
+// getContent returns all the fields of the entry as a json-string.
+func (t *Tailer) getContent(entry *sdjournal.JournalEntry) []byte {
+	// delete the hostname from the entry fields to ensure the hostname computed by the agent won't be overridden later on
+	fields := entry.Fields
+	delete(fields, sdjournal.SD_JOURNAL_FIELD_HOSTNAME)
+
+	var payload map[string]string
+	if !t.source.Config.DisableNormalization {
```

Comment:
Why do we need this? Could you add comments about real world use cases where we would normalize and where we would not normalize?

Reply:
Much like https://github.com/reevoo/fluent-plugin-systemd#filter-plugin-configuration, as a user, I feel having sane defaults (normalize fields enabled), but still maintain the ability to disable that behaviour (understanding the consequences) is a good way to:

* be explicit about the fact that the agent changes the content that comes out of `journald`
* let the power-user disable this if he feels this is incompatible with his setup.

Reply:
I'd rather remove this feature, I am not sure we really need it.